File: Programmers/[67258] 보석 쇼핑/woong-jae/67258.js
function solution(gems) {
    let answer = [0, gems.length - 1];
    const kind = new Set(gems).size;

    let start = 0, cur = 0, picked = new Set();
    while(start < gems.length) {
        while(cur < gems.length && picked.size < kind) {
            picked.add(gems[cur++]);
        }
        if(picked.size === kind) {
            let end = --cur;
            picked.clear();
            while(picked.size < kind) picked.add(gems[cur--]);
            start = cur + 1;
            if(answer[1] - answer[0] > end - start) {
                answer = [start, end];
                if(end - start + 1 === kind) break;
            }
            picked.delete(gems[start++]);
            cur = end + 1;
        }
        else break;
    }

    return answer.map(i => i + 1);
}

File: Programmers/[67258] 보석 쇼핑/Johoseong/67258.py
def solution(gems):
    answer = []
    gem_dict = dict()
    gem_unique = len(set(gems))
    start = 0
    end = 0
    section = len(gems) + 1 # 결과 진열대 범위 (gems 길이보다 길 순 없으니까 이렇게 초기화)

    while end < len(gems):
        if gems[end] not in gem_dict: # 발견 안된 보석이면 딕셔너리에 추가
            gem_dict[gems[end]] = 1
        else: # 이미 발견한 적 있는거면 발견 횟수 증가
            gem_dict[gems[end]] += 1
        
        end += 1 # 끝점 한칸 이동

        if len(gem_dict) == gem_unique: # 보석 종류별로 다 찾았으면?
            while start < end: # 더 짧은 범위인지 검사함
                if gem_dict[gems[start]] > 1: # 한번보다 더 발견됐다 = 뒤에 더 있다
                    gem_dict[gems[start]] -= 1 # 시작점 이동할 거니까 발견 값 하나 빼줌
                    start += 1 # 시작점 한칸 이동
                elif (end - start) < section: # 이렇게 갱신한 범위가 이전에 찾은 범위보다 작으면?
                    section = end - start 
                    answer = [start + 1, end] # 범위 갱신
                    break
                else: break

    return answer